File: FitoGrano/WinForm/frSecurityUsers.vb
Imports System.Data
Imports DevExpress.Utils
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo

Public Class frSecurityUsers
    Friend NewRecord_PerformeId As String

    ' закрытие формы
    Private Sub frSecurityUsers_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing
        clsMyUserFunction.RemoveTableDataSet(Me)
    End Sub
    ' загрузка формы
    Private Sub frSecurityUsers_Load(sender As System.Object, e As System.EventArgs) Handles MyBase.Load
        Me.lbServerName.Caption = pref_ServerIP
        Me.lbDB_name.Caption = pref_DataBaseName
        DesignerPerformers()
        'gvPerformers_RowCellClick(Me.gvPerformers, Nothing)
    End Sub

#Region "Дизайнеры"
    ' дизайнер исполнителей
    Private Sub DesignerPerformers()
        EventChangedControl = False
        SelectQueryData("Performers." & Me.Name, "EXEC PerformersFunctions @Function = 1")
        If Me.gcPerformers.DataSource Is Nothing Then
            Me.gcPerformers.DataSource = iDataSet.Tables("Performers." & Me.Name)
            mdHidenAllColumns.HidenAllColumns_Grid(Me.gvPerformers, iDataSet.Tables("Performers." & Me.Name))
            With Me.gvPerformers
                .Columns("RolesJobShort").Visible = True
                .Columns("RolesJobShort").Caption = "Должность"
                .Columns("SNP").Visible = True
                .Columns("SNP").Caption = "ФИО"
            End With
        End If
        EventChangedControl = True
        Me.gvPerformers.BestFitColumns()
        Me.gvPerformers.FocusedRowHandle = Me.gvPerformers.LocateByValue("PerformerId", CType(NewRecord_PerformeId, Integer))
    End Sub
    ' дизайнер контактов пользователя
    Private Sub DesignerPerformerContacts(ByRef PerformerId As String)
        PerformerId = clsForSQL.InDataBase(PerformerId)
        SelectQueryData(
                            "PerformersContacts." & Me.Name,
                            "EXEC PerformersContactsFunction @PerformerId = " & PerformerId & ", @Function = 1",
                            "GetPerformersContacts"
                            )
        Me.lbxContacts.DataSource = iDataSet.Tables("PerformersContacts." & Me.Name)
        Me.lbxContacts.DisplayMember = "NumberString"
    End Sub
    ' Дизайнер групп пользователей 
    Private Sub DesignerPerfGroups(ByRef PerformerId As String)
        PerformerId = clsForSQL.InDataBase(PerformerId)
        SelectQueryData(
                        "PerformersRightGroups." & Me.Name,
                        "SELECT * FROM vPerformersGroups pg 
                         WHERE pg.PerformerId = " & PerformerId,
                        "GetPerformersRightGroups"
                        )
        Me.lbxPerfGroups.DataSource = iDataSet.Tables("PerformersRightGroups." & Me.Name)
        Me.lbxPerfGroups.DisplayMember = "Name"
        Me.lbxPerfGroups.ValueMember = "Id"
    End Sub
#End Region

#Region "События gvPerformers"
    ' изменение фокуса в таблиуе исполнителей
    Private Sub gvPerformers_FocusedRowChanged(sender As Object, e As DevExpress.XtraGrid.Views.Base.FocusedRowChangedEventArgs) Handles gvPerformers.FocusedRowChanged
        Dim iRow As DataRow = sender.GetFocusedDataRow
        If EventChangedControl Then NewRecord_PerformeId = iRow("PerformerId").ToString
        Me.txtAddress.Text = iRow("Address").ToString
        Me.gcPerformerInformation.Text = " Информация о сотруднике <u>" & iRow("SNP").ToString & "</u>"
        Me.txtDtBirth.Text = If(iRow("DtBirth").ToString <> "", CType(iRow("DtBirth").ToString, Date).ToShortDateString, "")
        Me.txtRoleJob.Text = iRow("RolesJobShort").ToString
        Me.txtRoomNumber.Text = iRow("RoomNumber").ToString
        Me.txtSex.Text = iRow("Sex").ToString
        Me.txtSex.Properties.ContextImageOptions.Image = If(iRow("SexId").ToString = 0, My.Resources.SexWomen_16x16, My.Resources.SexMen_16x16)
        Me.txtEmail.Text = iRow("Email").ToString
        DesignerPerformerContacts(iRow("PerformerId").ToString)
        DesignerPerfGroups(iRow("PerformerId").ToString)
    End Sub
    ' двойной клик по исполнителю
    Private Sub gvPerformers_RowCellClick(sender As Object, e As DevExpress.XtraGrid.Views.Grid.RowCellClickEventArgs) Handles gvPerformers.RowCellClick
        If e.Clicks = 2 And e.Button = Windows.Forms.MouseButtons.Left And btnEditPerformer.Enabled Then
            btnEditPerformer_ItemClick(btnEditPerformer, Nothing)
        End If
        If e.Clicks = 1 Then
            gvPerformers_FocusedRowChanged(gvPerformers, Nothing)
        End If
    End Sub
    ' Перехват кнопки DELETE на исполнителях
    Private Sub gvPerformers_KeyDown(sender As Object, e As KeyEventArgs) Handles gvPerformers.KeyDown
        If e.KeyData = Keys.Delete Then
            btnDelPerformer_ItemClick(btnDelPerformer, Nothing)
        End If
    End Sub
    ' вызов контекстного меню на исполнителях
    Private Sub gvPerformers_MouseDown(sender As Object, e As MouseEventArgs) Handles gvPerformers.MouseDown
        Dim view As GridView = CType(sender, GridView)
        Dim pt As Point = view.GridControl.PointToClient(Control.MousePosition)
        Dim info As GridHitInfo = view.CalcHitInfo(pt)
        If e.Button = MouseButtons.Right And e.Clicks = 1 Then
            Me.btnEditPerformer_pm.Enabled = info.InRow
            Me.btnDelPerformer_pm.Enabled = info.InRow
            Me.pmMenuPerformers.ShowPopup(MousePosition)
        End If
    End Sub
#End Region

#Region "Управление исполнителями"
    ' добавить исполнителя
    Private Sub btnAddPerformer_ItemClick(sender As System.Object, e As DevExpress.XtraBars.ItemClickEventArgs) Handles btnAddPerformer.ItemClick, btnAddPerformer_pm.ItemClick
        frAddEditPerformer.AddOrEdit = 2
        If frAddEditPerformer.ShowDialog() = DialogResult.OK Then
            DesignerPerformers()
        End If
    End Sub
    ' редиктировать исполнителя
    Private Sub btnEditPerformer_ItemClick(sender As Object, e As DevExpress.XtraBars.ItemClickEventArgs) Handles btnEditPerformer.ItemClick, btnEditPerformer_pm.ItemClick
        frAddEditPerformer.AddOrEdit = 3
        If frAddEditPerformer.ShowDialog = DialogResult.OK Then
            DesignerPerformers()
        End If
    End Sub
    ' удалить исполнителя
    Private Sub btnDelPerformer_ItemClick(sender As Object, e As DevExpress.XtraBars.ItemClickEventArgs) Handles btnDelPerformer.ItemClick, btnDelPerformer_pm.ItemClick
        If XtraMessageBox.Show("Пользователь <b><u>" & Me.gvPerformers.GetFocusedRowCellDisplayText("SNP") & "</u></b> будет удален." & Chr(10) &
                               "Вы согласны?", Application.ProductName, MessageBoxButtons.YesNo, MessageBoxIcon.Question, DefaultBoolean.True) = DialogResult.Yes Then
            If ExecuteScalar(
                             "EXEC PerformersFunctions 
                                @Login = " & Me.gvPerformers.GetFocusedRowCellDisplayText("Login") & ", 
                                @Function = 4",
                             "DeleteUser"
                             ) <> "err" Then
                DesignerPerformers()
            End If
        End If
    End Sub
#End Region

    ' контектные кнопки информации о сотруднике
    Private Sub gcPerformerInformation_CustomButtonClick(sender As System.Object, e As DevExpress.XtraBars.Docking2010.BaseButtonEventArgs) Handles gcPerformerInformation.CustomButtonClick
        Dim _min As Integer = 30
        Dim _max As Integer = 193
        Select Case e.Button.Properties.GroupIndex
            Case 0 ' Статистика подключений

            Case 2 ' Свернуть / Развернуть
                Select Case e.Button.Properties.Tag
                    Case 0 ' свернуть
                        e.Button.Properties.Tag = 1
                        e.Button.Properties.ImageOptions.Image = My.Resources.down_20x20
                        Me.gcPerformerInformation.Size = New Size(Me.gcPerformerInformation.Size.Width, 30)
                        Me.tlRightsPerformer.Location = New Point(Me.tlRightsPerformer.Location.X, Me.tlRightsPerformer.Location.Y - (_max - _min))
                        Me.tlRightsPerformer.Size = New Point(Me.tlRightsPerformer.Size.Width, Me.tlRightsPerformer.Size.Height + (_max - _min))
                    Case 1 ' развернуть
                        e.Button.Properties.Tag = 0
                        e.Button.Properties.ImageOptions.Image = My.Resources.up_20x20
                        Me.gcPerformerInformation.Size = New Size(Me.gcPerformerInformation.Size.Width, 193)
                        Me.tlRightsPerformer.Location = New Point(Me.tlRightsPerformer.Location.X, Me.tlRightsPerformer.Location.Y + (_max - _min))
                        Me.tlRightsPerformer.Size = New Point(Me.tlRightsPerformer.Size.Width, Me.tlRightsPerformer.Size.Width - (_max - _min))
                End Select
        End Select
    End Sub
    ' таймер для активности кнопок
    Private Sub tmChecked_Tick(sender As Object, e As EventArgs) Handles tmChecked.Tick
        Dim gvPerformersRowCount As Integer = Me.gvPerformers.RowCount

        SplitContainerControl1.Panel2.Enabled = gvPerformersRowCount <> 0
        Me.btnEditPerformer.Enabled = gvPerformersRowCount <> 0
        Me.btnDelPerformer.Enabled = gvPerformersRowCount <> 0
    End Sub
    ' обновить форму
    Private Sub btnUpdate_ItemClick(sender As Object, e As DevExpress.XtraBars.ItemClickEventArgs) Handles btnUpdate.ItemClick
        DesignerPerformers()
    End Sub
End Class

File: FitoGrano/CS/WinForm/frSecurityUsers.cs
// VBConversions Note: VB project level imports
using System.Collections.Generic;
using System;
using System.Linq;
using System.Drawing;
using System.Diagnostics;
using Microsoft.VisualBasic;
using System.Xml.Linq;
using System.Windows.Forms;
// End of VB project level imports

using System.Data;
using DevExpress.Utils;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using FitoGrano;

namespace FitoGrano
{
	
	public partial class frSecurityUsers
	{
		public frSecurityUsers()
		{
			InitializeComponent();
			
			//Added to support default instance behavour in C#
			if (defaultInstance == null)
				defaultInstance = this;
		}
		
#region Default Instance
		
		private static frSecurityUsers defaultInstance;
		
		/// <summary>
		/// Added by the VB.Net to C# Converter to support default instance behavour in C#
		/// </summary>
		public static frSecurityUsers Default
		{
			get
			{
				if (defaultInstance == null)
				{
					defaultInstance = new frSecurityUsers();
					defaultInstance.FormClosed += new System.Windows.Forms.FormClosedEventHandler(defaultInstance_FormClosed);
				}
				
				return defaultInstance;
			}
			set
			{
				defaultInstance = value;
			}
		}
		
		static void defaultInstance_FormClosed(object sender, System.Windows.Forms.FormClosedEventArgs e)
		{
			defaultInstance = null;
		}
		
#endregion
		internal string NewRecord_PerformeId;
		
		// закрытие формы
		public void frSecurityUsers_FormClosing(object sender, FormClosingEventArgs e)
		{
			clsMyUserFunction.RemoveTableDataSet(this);
		}
		// загрузка формы
		public void frSecurityUsers_Load(System.Object sender, System.EventArgs e)
		{
			this.lbServerName.Caption = mdVariables.pref_ServerIP;
			this.lbDB_name.Caption = mdVariables.pref_DataBaseName;
			DesignerPerformers();
			//gvPerformers_RowCellClick(Me.gvPerformers, Nothing)
		}
		
#region Дизайнеры
		// дизайнер исполнителей
		private void DesignerPerformers()
		{
			mdVariables.EventChangedControl = false;
			mdQuery.SelectQueryData("Performers." + this.Name, "EXEC PerformersFunctions @Function = 1");
			if (ReferenceEquals(this.gcPerformers.DataSource, null))
			{
				this.gcPerformers.DataSource = mdQuery.iDataSet.Tables["Performers." + this.Name];
				mdHidenAllColumns.HidenAllColumns_Grid(this.gvPerformers, mdQuery.iDataSet.Tables["Performers." + this.Name]);
				this.gvPerformers.Columns["RolesJobShort"].Visible = true;
				this.gvPerformers.Columns["RolesJobShort"].Caption = "Должность";
				this.gvPerformers.Columns["SNP"].Visible = true;
				this.gvPerformers.Columns["SNP"].Caption = "ФИО";
			}
			mdVariables.EventChangedControl = true;
			this.gvPerformers.BestFitColumns();
			this.gvPerformers.FocusedRowHandle = this.gvPerformers.LocateByValue("PerformerId", System.Convert.ToInt32(NewRecord_PerformeId));
		}
		// дизайнер контактов пользователя
		private void DesignerPerformerContacts(ref string PerformerId)
		{
			PerformerId = clsForSQL.InDataBase(PerformerId);
			mdQuery.SelectQueryData(
				"PerformersContacts." + this.Name,
				"EXEC PerformersContactsFunction @PerformerId = " + PerformerId + ", @Function = 1",
				"GetPerformersContacts"
				);
			this.lbxContacts.DataSource = mdQuery.iDataSet.Tables["PerformersContacts." + this.Name];
			this.lbxContacts.DisplayMember = "NumberString";
		}
		// Дизайнер групп пользователей
		private void DesignerPerfGroups(ref string PerformerId)
		{
			PerformerId = clsForSQL.InDataBase(PerformerId);
			mdQuery.SelectQueryData(
				"PerformersRightGroups." + this.Name, System.Convert.ToString(
				SELECT * FROM vPerformersGroups pg
				WHERE pg.PerformerId == & PerformerId,
				"GetPerformersRightGroups")
				);
			this.lbxPerfGroups.DataSource = mdQuery.iDataSet.Tables["PerformersRightGroups." + this.Name];
			this.lbxPerfGroups.DisplayMember = "Name";
			this.lbxPerfGroups.ValueMember = "Id";
		}
#endregion
		
#region События gvPerformers
		// изменение фокуса в таблиуе исполнителей
		public void gvPerformers_FocusedRowChanged(dynamic sender, DevExpress.XtraGrid.Views.Base.FocusedRowChangedEventArgs e)
		{
			DataRow iRow = sender.GetFocusedDataRow;
			if (mdVariables.EventChangedControl)
			{
				NewRecord_PerformeId = iRow["PerformerId"].ToString();
			}
			this.txtAddress.Text = iRow["Address"].ToString();
			this.gcPerformerInformation.Text = " Информация о сотруднике <u>" + iRow["SNP"].ToString() + "</u>";
			this.txtDtBirth.Text = System.Convert.ToString((!string.IsNullOrEmpty(iRow["DtBirth"].ToString())) ? (System.Convert.ToDateTime(iRow["DtBirth"].ToString()).ToShortDateString()) : "");
			this.txtRoleJob.Text = iRow["RolesJobShort"].ToString();
			this.txtRoomNumber.Text = iRow["RoomNumber"].ToString();
			this.txtSex.Text = iRow["Sex"].ToString();
			this.txtSex.Properties.ContextImageOptions.Image = (iRow["SexId"].ToString() == 0) ? global::My.Resources.Resources.SexWomen_16x16 : global::My.Resources.Resources.SexMen_16x16;
			this.txtEmail.Text = iRow["Email"].ToString();
			string temp_PerformerId = iRow["PerformerId"].ToString();
			DesignerPerformerContacts(ref temp_PerformerId);
			string temp_PerformerId2 = iRow["PerformerId"].ToString();
			DesignerPerfGroups(ref temp_PerformerId2);
		}
		// двойной клик по исполнителю
		public void gvPerformers_RowCellClick(object sender, DevExpress.XtraGrid.Views.Grid.RowCellClickEventArgs e)
		{
			if (e.Clicks == 2 & e.Button == System.Windows.Forms.MouseButtons.Left && btnEditPerformer.Enabled)
			{
				btnEditPerformer_ItemClick(btnEditPerformer, null);
			}
			if (e.Clicks == 1)
			{
				gvPerformers_FocusedRowChanged(gvPerformers, null);
			}
		}
		// Перехват кнопки DELETE на исполнителях
		public void gvPerformers_KeyDown(object sender, KeyEventArgs e)
		{
			if (e.KeyData == Keys.Delete)
			{
				btnDelPerformer_ItemClick(btnDelPerformer, null);
			}
		}
		// вызов контекстного меню на исполнителях
		public void gvPerformers_MouseDown(object sender, MouseEventArgs e)
		{
			GridView view = (GridView) sender;
			Point pt = view.GridControl.PointToClient(Control.MousePosition);
			GridHitInfo info = view.CalcHitInfo(pt);
			if (e.Button == MouseButtons.Right & e.Clicks == 1)
			{
				this.btnEditPerformer_pm.Enabled = info.InRow;
				this.btnDelPerformer_pm.Enabled = info.InRow;
				this.pmMenuPerformers.ShowPopup(MousePosition);
			}
		}
#endregion
		
#region Управление исполнителями
		// добавить исполнителя
		public void btnAddPerformer_ItemClick(System.Object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{
			frAddEditPerformer.Default.AddOrEdit = 2;
			if (frAddEditPerformer.Default.ShowDialog() == DialogResult.OK)
			{
				DesignerPerformers();
			}
		}
		// редиктировать исполнителя
		public void btnEditPerformer_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{
			frAddEditPerformer.Default.AddOrEdit = 3;
			if (frAddEditPerformer.Default.ShowDialog() == DialogResult.OK)
			{
				DesignerPerformers();
			}
		}
		// удалить исполнителя
		public void btnDelPerformer_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{
			if (XtraMessageBox.Show("Пользователь <b><u>" + this.gvPerformers.GetFocusedRowCellDisplayText("SNP") + "</u></b> будет удален." + "\n" +
					"Вы согласны?", Application.ProductName, MessageBoxButtons.YesNo, MessageBoxIcon.Question, DefaultBoolean.True) == DialogResult.Yes)
					{
					if (ExecuteScalar(
							EXEC PerformersFunctions
							@Login == " & Me.gvPerformers.GetFocusedRowCellDisplayText(" Login ") & ",
							@Function == 4,
							"DeleteUser"
							) != "err")
							{
							DesignerPerformers();
				}
			}
		}
#endregion
		
		// контектные кнопки информации о сотруднике
		public void gcPerformerInformation_CustomButtonClick(System.Object sender, DevExpress.XtraBars.Docking2010.BaseButtonEventArgs e)
		{
			int _min = 30;
			int _max = 193;
			switch (e.Button.Properties.GroupIndex)
			{
				case 0: // Статистика подключений
					break;
					
				case 2: // Свернуть / Развернуть
					if ((int) e.Button.Properties.Tag == 0) // свернуть
					{
						e.Button.Properties.Tag = 1;
						e.Button.Properties.ImageOptions.Image = global::My.Resources.Resources.down_20x20;
						this.gcPerformerInformation.Size = new Size(this.gcPerformerInformation.Size.Width, 30);
						this.tlRightsPerformer.Location = new Point(this.tlRightsPerformer.Location.X, this.tlRightsPerformer.Location.Y - (_max - _min));
						this.tlRightsPerformer.Size = new Point(this.tlRightsPerformer.Size.Width, this.tlRightsPerformer.Size.Height + (_max - _min));
					} // развернуть
					else if ((int) e.Button.Properties.Tag == 1)
					{
						e.Button.Properties.Tag = 0;
						e.Button.Properties.ImageOptions.Image = global::My.Resources.Resources.up_20x20;
						this.gcPerformerInformation.Size = new Size(this.gcPerformerInformation.Size.Width, 193);
						this.tlRightsPerformer.Location = new Point(this.tlRightsPerformer.Location.X, this.tlRightsPerformer.Location.Y + (_max - _min));
						this.tlRightsPerformer.Size = new Point(this.tlRightsPerformer.Size.Width, this.tlRightsPerformer.Size.Width - (_max - _min));
					}
					break;
			}
		}
		// таймер для активности кнопок
		public void tmChecked_Tick(object sender, EventArgs e)
		{
			int gvPerformersRowCount = this.gvPerformers.RowCount;
			
			SplitContainerControl1.Panel2.Enabled = gvPerformersRowCount != 0;
			this.btnEditPerformer.Enabled = gvPerformersRowCount != 0;
			this.btnDelPerformer.Enabled = gvPerformersRowCount != 0;
		}
		// обновить форму
		public void btnUpdate_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
		{
			DesignerPerformers();
		}
	}
}
